examples/keras_io/vision/metric_learning.py

```diff
@@ -0,0 +1,323 @@
+    loss=keras.losses.SparseCategoricalCrossentropy(from_logits=True),
+)
+
+# history = model.fit(AnchorPositivePairs(num_batches=1000), epochs=20)
+history = model.fit(AnchorPositivePairs(num_batches=1000), epochs=1)
+
+plt.plot(history.history["loss"])
+plt.show()
+
+"""
+## Testing
+
+We can review the quality of this model by applying it to the test set and considering
+near neighbours in the embedding space.
+
+First we embed the test set and calculate all near neighbours. Recall that since the
+embeddings are unit length we can calculate cosine similarity via dot products.
+"""
+
+near_neighbours_per_example = 10
+
+embeddings = model.predict(x_test)
+gram_matrix = np.einsum("ae,be->ab", embeddings, embeddings)
+near_neighbours = np.argsort(gram_matrix.T)[:, -(near_neighbours_per_example + 1) :]
+
+"""
+As a visual check of these embeddings we can build a collage of the near neighbours for 5
+random examples. The first column of the image below is a randomly selected image, the
+following 10 columns show the nearest neighbours in order of similarity.
+"""
+
+num_collage_examples = 5
+
+examples = np.empty(
+    (
+        num_collage_examples,
+        near_neighbours_per_example + 1,
+        height_width,
+        height_width,
+        3,
+    ),
+    dtype=np.float32,
+)
+for row_idx in range(num_collage_examples):
+    examples[row_idx, 0] = x_test[row_idx]
+    anchor_near_neighbours = reversed(near_neighbours[row_idx][:-1])
+    for col_idx, nn_idx in enumerate(anchor_near_neighbours):
+        examples[row_idx, col_idx + 1] = x_test[nn_idx]
+
+show_collage(examples)
+
+"""
+We can also get a quantified view of the performance by considering the correctness of
+near neighbours in terms of a confusion matrix.
+
+Let us sample 10 examples from each of the 10 classes and consider their near neighbours
+as a form of prediction; that is, does the example and its near neighbours share the same
+class?
+
+We observe that each animal class does generally well, and is confused the most with the
+other animal classes. The vehicle classes follow the same pattern.
+"""
+
+confusion_matrix = np.zeros((num_classes, num_classes))
+
+# For each class.
+for class_idx in range(num_classes):
+    # Consider 10 examples.
+    example_idxs = class_idx_to_test_idxs[class_idx][:10]
+    for y_test_idx in example_idxs:
+        # And count the classes of its near neighbours.
+        for nn_idx in near_neighbours[y_test_idx][:-1]:
+            nn_class_idx = y_test[nn_idx]
+            confusion_matrix[class_idx, nn_class_idx] += 1
+
+# Display a confusion matrix.
+labels = [
+    "Airplane",
+    "Automobile",
+    "Bird",
+    "Cat",
+    "Deer",
+    "Dog",
+    "Frog",
+    "Horse",
+    "Ship",
+    "Truck",
+]
+disp = ConfusionMatrixDisplay(confusion_matrix=confusion_matrix, display_labels=labels)
+disp.plot(include_values=True, cmap="viridis", ax=None, xticks_rotation="vertical")
+plt.show()
+
+"""
+Example available on HuggingFace.
```

Comment:
remove this section about hugging face and the table

Reply:
done